# Week 2 Exercises
Approximations and round-off errors   
Luong Nguyen, 31.1.2017

In [68]:
import numpy as np

## Exercise 3

In [69]:
x = 0.0
eps = 1.0
while (x + eps > x):
    eps = eps/2
eps*2

0.0

In [70]:
x = 0.0
eps = 0.1
res = []
while (x + eps > x):
    eps = eps/2
    res.append(eps)
res[-2]

5e-324

In [71]:
5e-324 + 0.0 > 0.0

True

In [72]:
np.nextafter(0,1)

4.9406564584124654e-324

In [73]:
4.9406564584124654e-324 + 0.0 > 0.0

True

# Exercise 4

In [74]:
x = 1000000000000000.0
eps = 1.0
while (x + eps > x):
    eps = eps/2
eps*2

0.125

When the study point increases, the eps also increases. They have positive relationship.

# Exercise 5

In [75]:
x = 0.577
df = 6 * x / (1 - 3 * x**2)**2
df

2352910.792601485

In [76]:
from decimal import *
for i in range(3, 17):
    getcontext().prec = i
    x = Decimal('0.577')
    df = 6 * x / (1 - 3 * x**2)**2
    print(i, "significant digits - df =", df)

3 significant digits - df = 3.46E+6
4 significant digits - df = 2.049E+6
5 significant digits - df = 2.3646E+6
6 significant digits - df = 2.35291E+6
7 significant digits - df = 2352911
8 significant digits - df = 2352910.8
9 significant digits - df = 2352910.79
10 significant digits - df = 2352910.793
11 significant digits - df = 2352910.7926
12 significant digits - df = 2352910.79260
13 significant digits - df = 2352910.792602
14 significant digits - df = 2352910.7926020
15 significant digits - df = 2352910.79260199
16 significant digits - df = 2352910.792601992


As the number of significant figures increases, the value of the derivative converges to 2352910.7926 